### PR TITLE
Pin char when UTF-16 marshalling is specified and runtime marshalling is disabled

### PR DESCRIPTION
**Pull request by elinor-fung**
Contributes to https://github.com/dotnet/runtime/issues/69809: item (2)

When `DisableRuntimeMarshalling` is used and marshalling info (`StringMarshalling.Utf16` or `UnmanagedType.U2/I2`) is specified for `char`, we were treating it as if runtime marshalling was not disabled, so we wouldn't pin.

```csharp
[assembly: DisableRuntimeMarshalling]

[LibraryImport("lib", StringMarshalling = StringMarshalling.Utf16)]
public static partial void CharArray(char[] c);
```
Old generated code:
```csharp
byte* __c_native = default;
// Setup - Perform required setup.
global::System.Runtime.InteropServices.Marshalling.ArrayMarshaller<char> __c_native__marshaller = default;
try
{
    // Marshal - Convert managed data to native data.
    byte* __c_native__marshaller__stackptr = stackalloc byte[512];
    __c_native__marshaller = new(c, new System.Span<byte>(__c_native__marshaller__stackptr, 512), sizeof(ushort));
    __c_native__marshaller.GetManagedValuesSource().CopyTo(System.Runtime.InteropServices.MemoryMarshal.Cast<byte, char>(__c_native__marshaller.GetNativeValuesDestination()));
    // Pin - Pin data in preparation for calling the P/Invoke.
    fixed (void* __c_native__ignored = __c_native__marshaller)
    {
        __c_native = __c_native__marshaller.ToNativeValue();
        __PInvoke(__c_native);
    }
}
finally
{
    // Cleanup - Perform required cleanup.
    __c_native__marshaller.FreeNative();
}
```
New generated code:
```csharp
// Marshal - Convert managed data to native data.
ref char __byref_c = ref (c == null ? ref *(char*)0 : ref System.Runtime.InteropServices.MemoryMarshal.GetArrayDataReference(c));
// Pin - Pin data in preparation for calling the P/Invoke.
fixed (byte* __c_native = &System.Runtime.CompilerServices.Unsafe.As<char, byte>(ref __byref_c))
{
    __PInvoke(__c_native);
}
```

### CODE REVIEW

File: src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/InteropGenerationOptions.cs [hunk 1]
```diff
@@ -7,5 +7,5 @@
 
 namespace Microsoft.Interop
 {
-    public readonly record struct InteropGenerationOptions(bool UseMarshalType);
+    public readonly record struct InteropGenerationOptions(bool UseMarshalType, bool RuntimeMarshallingDisabled);
```
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/InteropGenerationOptions.cs:10]: I think this new information would fit better on the `StubCodeContext`.

Comment (elinor-fung) [reply src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/InteropGenerationOptions.cs:10]: My thought was that we currently don't need that information beyond creating the generator, so I kept it off the `StubCodeContext` and put it on the options passed to the generator factory - similar to `AttributedMarshallingModelGeneratorFactory`.

**Review by AaronRobinsonMSFT — APPROVED** [review]

**Review by AaronRobinsonMSFT — APPROVED** [review]

**Review by AaronRobinsonMSFT — APPROVED** [review]

**Comment by elinor-fung** [comment]
Failure is real - https://github.com/dotnet/runtime/pull/69914 should fix it.

**Comment by jkotas** [comment]
> `fixed (byte* __c_native = &System.Runtime.CompilerServices.Unsafe.As<char, byte>(ref __byref_c))`

Can the type be changed to `void*` to make the `Unsafe.As` cast unnecessary?

**Pull request merged**